Add overlay options to Tado

north3221 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I have added the ability to set the overlay mode, this is both in the service for setting the climate timer and also in the integration options. This way you can decide default overlay type and also override that default when calling the service.

This is useful for things like google assistant and manual changes to temp in the HA app. Meaning you can set it to TADO_DEFAULT and have the integration manage the changes based on whatever settings you have as default in the Tado app. Yet still override that in an automation should you choose by overriding that setting in the service call

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: Adding to the set climate timer service as part of code pr #21512

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

north3221 committed:
Added the ability to choose what overlay type to use by default and on service call

north3221 committed:
Cleaned up a few bits left over

north3221 committed:
Added config migration and matched defaults to previous version

north3221 committed:
no longer need const manual in config entry

michaelarnauts commented:
@north3221 would you mind taking over the ownership of the tado integration? I have so little spare time lately to look into this.

north3221 committed:
Adressing run failures and fied update zones to use zones recieved

north3221 committed:
Adressing run failures

north3221 commented:
> @north3221 would you mind taking over the ownership of the tado integration? I have so little spare time lately to look into this.

I don't feel qualified 🤣 

Although given I have a couple of pr's to raises, makes sense I say yes then if someone else starts doing more pr's can look to pass the mantle then...

north3221 committed:
Adressing run failures

north3221 committed:
Adressing test failures

north3221 committed:
Re ran black for tests

north3221 committed:
Adressing test failures

north3221 committed:
Changed owner at @michaelarnauts request

north3221 committed:
Updated owner at @michaelarnauts request

north3221 committed:
Revert "Updated CODEOWNERS file incorrectly"

This reverts commit 47c2a23a6b5d8a84670224945e01f49a0a14cf54.

north3221 committed:
Trying updating codeowner again, have added hassfest to my git...

north3221 committed:
retrigger checks

north3221 committed:
fixed codeowners?

north3221 commented:
@michaelarnauts ok I've updated everything for this PR so all test pass.. and added myself to codeowners as requested... think I got it all set up properly now..

north3221 commented:
@michaelarnauts I'm guessing you will need to approve this pr so it adds me in?

michaelarnauts commented:
sorry for the delay!

balloob commented on homeassistant/components/tado/__init__.py:
```diff
@@ -121,6 +123,29 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Migrate old entry."""
+    _LOGGER.debug("Migrating from version %s for %s", config_entry.version, DOMAIN)
+
+    if config_entry.version == 1:
```
We should avoid migrations if we can, and it looks like we can migrate this while fetching the data from the options.

Config entry migrations will prevent the user from rolling back Home Assistant versions so should be avoided unless really necessary.

north3221 replied:
Sorry for the delay, not had any spare time.

Ahh ok, yeah that makes sense, I thought I was doing the right thing with migration flow, but yeah no worries, I'll take it out

balloob commented on homeassistant/components/tado/config_flow.py:
```diff
@@ -11,12 +11,24 @@
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+        vol.Optional(CONF_FALLBACK, default=CONST_OVERLAY_TADO_MODE): vol.In(
```
Don't add this to the config flow. The config flow should only gather info needed to authenticate.

This should go into an options flow.

balloob commented on homeassistant/components/tado/services.yaml:
```diff
@@ -24,6 +16,26 @@ set_climate_timer:
+      default: "01:00:00"
+      selector:
+        text:
+    requested_overlay:
```
It's weird that a service called `set_climate_timer` allows setting a `requested_overlay` which is not compatible with setting a time period.

north3221 replied:
Yeah I get what you mean, but the reality is that set_climate_timer is actually setting an overlay in tado, all I am adding is the ability to choose which type of overlay. Probably the service should of been called set_overlay, but I dont want to change that due to backwards compatibility

balloob commented on homeassistant/components/tado/services.yaml:
```diff
@@ -24,6 +16,26 @@ set_climate_timer:
+      selector:
+        select:
+          options:
+            - "NEXT_TIME_BLOCK"
```
Can we make these options a bit more descriptive/friendly?

balloob commented on homeassistant/components/tado/climate.py:
```diff
@@ -58,12 +62,14 @@
         cv.time_period, cv.positive_timedelta, lambda td: td.total_seconds()
     ),
-    vol.Required(ATTR_TEMPERATURE): vol.Coerce(float),
+    vol.Optional(ATTR_REQUESTED_OVERLAY): vol.In(CONST_OVERLAY_TADO_OPTIONS),
```
If we keep these 2 options but they shouldn't be combined, use `vol.Exclusive`

north3221 committed:
Addressing feedback on pr

north3221 committed:
fix tests

north3221 commented:
@balloob I've looked to address all your feedback (thanks for taking the time and providing), think this is good to go now.

balloob committed:
Merge remote-tracking branch 'origin/dev' into add-overlay-options
